packages/upload-api/src/blob/add.js

```diff
@@ -0,0 +1,70 @@
+import * as Server from '@ucanto/server'
+import * as Blob from '@web3-storage/capabilities/blob'
+import * as API from '../types.js'
+
+/**
+ * @param {API.BlobServiceContext} context
+ * @returns {API.ServiceMethod<API.BlobAdd, API.BlobAddSuccess, API.BlobAddFailure>}
+ */
+export function blobAddProvider(context) {
+  return Server.provideAdvanced({
+    capability: Blob.add,
+    handler: async ({ capability }) => {
+      const { id, allocationStore, maxUploadSize } = context
+      const { content, size } = capability.nb
+      const space = /** @type {import('@ucanto/interface').DIDKey} */ (
+        Server.DID.parse(capability.with).did()
+      )
+
+      if (size > maxUploadSize) {
+        return {
+          error: new Server.Failure(
+            `Maximum size exceeded: ${maxUploadSize}, split DAG into smaller shards.`
+          ),
+        }
+      }
+      
+      // Create effects for receipt
+      const [allocatefx, acceptfx] = await Promise.all([
+        Blob.allocate
+          .invoke({
+            issuer: id,
+            audience: id,
+            with: id.toDIDKey(),
+            nb: {
+              content,
+              size,
+            },
+            expiration: Infinity,
+          })
+          .delegate(),
```

Comment:
I was thinking this one could be actual invocation that we run from here, I do not believe we need to schedule it for later. In my mind it would basically do what we have here


https://github.com/web3-storage/w3up/blob/9aba8a155cca663387952fc9b11bc3748de8f2fc/packages/upload-api/src/store/add.js#L13-L74


Than we would do what you do here. Followed by logic that checks if we already have blob, if so we could also issue receipt for accept. If we do not have it yet then do exactly what you do here.

Reply:
I'm not sure if we have a good way to check if we already have a receipt for this task or not, but assuming we can do that I imagine we'd have a big if clause around all these so if we do have a receipt we just check if blob was received and if so issue receipt for the `accept` if we do not have receipt yet it's a first run so we do what was described prior

Reply:
I just noticed this and also created #1342 with this branch with more iterations. I ported these comments over on https://github.com/web3-storage/w3up/pull/1342#pullrequestreview-1957849688